utils/billgeneration.js

```diff
@@ -1,20 +1,68 @@
 /* eslint-disable import/imports-first */
 import moment from 'moment';
 import dotenv from 'dotenv-safe';
+import fs from 'fs';
 
+import React from 'react';
+import ReactPDF from '@react-pdf/renderer';
 import {
   getSolarProjectById,
   getOwnerById,
   getRateScheduleById,
   createSubscriberBill,
   updateSubscriberBill,
-  getSubscriberBillsByStatementNumber
+  getSubscriberBillsByStatementNumber,
+  getSubscriberBillById
 } from '../airtable/request';
 import getEnphaseData from './enphase';
 import getLatestBill from './utilityApi';
+import BillingTemplate from './pdf/BillingTemplate';
+
+// const ROOT_URL = 'https://1f7c24d2.ngrok.io';
+const ROOT_URL = 'https://peoplepower-node.herokuapp.com';
 
 dotenv.config();
 
+const generatePdf = async (
+  subscriber,
+  solarProject,
+  subscriberBillId,
+  prevBillId
+) => {
+  // Refresh information about existing bills to get the latest calculated values
+  const subscriberBill = await getSubscriberBillById(subscriberBillId);
+  let prevBill;
+  if (prevBillId) {
+    prevBill = await getSubscriberBillById(prevBillId);
+  } else {
+    prevBill = {
+      amountDue: 0,
+      amountReceived: 0,
+      balance: 0
+    };
+  }
+  console.log(`Creating PDF for Bill# ${subscriberBill.id} ...`);
+  await ReactPDF.render(
+    <BillingTemplate
+      subscriber={subscriber}
+      solarProject={solarProject}
+      subscriberBill={subscriberBill}
+      prevBill={prevBill}
+    />,
+    `./temp/${subscriberBill.id}.pdf`
+  );
+  await updateSubscriberBill(subscriberBill.id, {
+    billPdf: [{ url: `${ROOT_URL}/${subscriberBill.id}.pdf` }]
+  });
+  console.log(
+    `Succesfully uploaded PDF at ${ROOT_URL}/${subscriberBill.id}.pdf`
+  );
+  setTimeout(() => {
+    console.log(`Deleting Temporary PDF: ${subscriberBill.id}.pdf`);
+    fs.unlinkSync(`./temp/${subscriberBill.id}.pdf`);
+  }, 20000);
```

Comment:
I wanna make sure I fully understand what's going on here: 

ReactPDF.render automatically saves a file to storage (which is in herokue, in this case). we then update the subscriber bill with a unique url? Where in the code is that pdf actually 'uploaded' to that url? 

After successfullyl doing so, we delete that pdf

Reply:
great question! 

so `ReactPDF.render` is saving it in the local folder called `temp`. In `index.js`, the line `app.use(express.static('temp'))` _exposes_ that `temp` folder to the web. Aka any files in that `temp` folder will be available at `peoplepower-node.herokuapp.com/filename`. Then, we give that url to airtable, it copies the pdf over, and we delete it once it's done